modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -814,10 +896,16 @@ void onRemovedInflights() {
             }
         }
 
-        void finishTx() {
+        void finishTx(Map<TablePartitionId, Long> enlistedGroups) {
+            this.enlistedGroups = enlistedGroups;
             finishInProgressFuture = new CompletableFuture<>();
         }
 
+        void setRollbackCause(Exception rollbackCause) {
+            this.rollbackCause = rollbackCause;
+            waitRepFut.complete(null);
```

Comment:
We would have to handle this in`TxManagerImpl.TxContext#performFinish` and call `finishAction` in the exceptional case, instead of simply calling it as a part of regular flow to abort this tx. Looks like too much of rework which doesn't make sense so far

Reply:
upd.: I did some refactoring as we discussed.